lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -1088,6 +1103,62 @@ def set_http_request(record, entry)
       entry.http_request = output
     end
 
+    def set_source_location(record, entry)
```

Comment:
Seems we are doing very similar things for `source_location`, `trace` and `operation`. Can we use some shared utility method to capture the logic and leave the dynamic part separate as a map. I'm thinking of something like this:

Map definition:
```
      # Map from each field name under LogEntry to corresponding variables
      # required to perform field value extraction from the log record.
      LOG_ENTRY_FIELDS_MAP = {
        source_location: [
          # The config to specify label name for field extraction from record.
          '@source_location_key',
          # Map from subfields' names to their types.
          {
            'file' => String.name,
            'function' => String.name,
            'line' => Numeric.name
          },
          # The grpc version class name.
          'Google::Logging::V2::LogEntrySourceLocation',
          # The non-grpc version class name.
          'Google::Apis::LoggingV2beta1::LogEntrySourceLocation'
        ],
        trace: ...,
        operation: ...
      }
```

Set fields in the main method:
```
    [:source_location, :trace, :operation].each do |field_name|
      set_log_entry_field(record, entry, field)
    end
```
Helper method:
```
    def set_log_entry_field(record, entry, field_name)
      field_name_config_variable, subfields_map, grpc_class_name,
        non_grpc_class_name = LOG_ENTRY_FIELDS_MAP[field_name]
      label_name = instance_variable_get(field_name_config_variable)
      subfields = fields_to_labels(record[label_name],
                                                    subfields_map.keys.map { |l| [l, l] }.to_h)
      record.delete(label_name) if record[label_name].empty?

      return unless subfields

      if @use_grpc
        output = Object.const_get(grpc_class_name).new
      else
        output = Object.const_get(non_grpc_class_name).new
      end
      subfields.each do |key, value|
        next unless value
        # Convert the string value to String or Integer objects.
        formatted_value = case subfields_map[key]
          when String.name
            value
          when Numeric.name
            value.to_i
          end
        output.send("#{key}=", formatted_value)
      end
      entry.send("#{field_name}=", output)
    rescue StandardError => err
      @log.error "Failed to set log entry field for #{field_name}.", err
    end
```